### PR TITLE
[Issue 7846]Add support for regex patterns in tenant and namespace

### LINKED ISSUE #7846: Multi-Topic subscription across tenant and namespace using regex in Java client throws IllegalArgumentException
**Describe the bug**
When using a regex to create a consumer and a regex pattern is present in the tenant or namespace section of the pattern this error is thrown. 

```
java.lang.IllegalArgumentException: Invalid named entity: \w+
	at org.apache.pulsar.common.naming.NamedEntity.checkName(NamedEntity.java:39)
	at org.apache.pulsar.common.naming.NamespaceName.validateNamespaceName(NamespaceName.java:179)
	at org.apache.pulsar.common.naming.NamespaceName.get(NamespaceName.java:53)
	at org.apache.pulsar.common.naming.TopicName.<init>(TopicName.java:151)
	at org.apache.pulsar.common.naming.TopicName.<init>(TopicName.java:38)
	at org.apache.pulsar.common.naming.TopicName$1.load(TopicName.java:63)
	at org.apache.pulsar.common.naming.TopicName$1.load(TopicName.java:60)
	at org.apache.pulsar.shade.com.google.common.cache.LocalCache$LoadingValueReference.loadFuture(LocalCache.java:3527)
	at org.apache.pulsar.shade.com.google.common.cache.LocalCache$Segment.loadSync(LocalCache.java:2276)
	at org.apache.pulsar.shade.com.google.common.cache.LocalCache$Segment.lockedGetOrLoad(LocalCache.java:2154)
	at org.apache.pulsar.shade.com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2044)
	at org.apache.pulsar.shade.com.google.common.cache.LocalCache.get(LocalCache.java:3951)
	at org.apache.pulsar.shade.com.google.common.cache.LocalCache.getOrLoad(LocalCache.java:3973)
	at org.apache.pulsar.shade.com.google.common.cache.LocalCache$LocalLoadingCache.get(LocalCache.java:4957)
	at org.apache.pulsar.common.naming.TopicName.get(TopicName.java:88)
	at org.apache.pulsar.client.impl.PulsarClientImpl.patternTopicSubscribeAsync(PulsarClientImpl.java:403)
	at org.apache.pulsar.client.impl.PulsarClientImpl.subscribeAsync(PulsarClientImpl.java:333)
	at org.apache.pulsar.client.impl.ConsumerBuilderImpl.subscribeAsync(ConsumerBuilderImpl.java:142)
	at org.apache.pulsar.client.impl.ConsumerBuilderImpl.subscribe(ConsumerBuilderImpl.java:99)
```

Example pattern: `persistent://\w+/Test/Test1` or `persistent://Test/\w+/Test1`

This is because `PulsarClientImpl.patternTopicSubscribeAsync` is called when creating a consumer using a Pattern. This calls `topicName.get(regex);` where regex is the Pattern passed in. This method attempts to validate the tenant and namespace names against this regex `^[-=:.\\w]*$` in org.apache.pulsar.common.naming.NamedEntity. This fails for any regex pattern. The namespace name is needed in the current code because it is used by `LookupService.getTopicsUnderNamespace` to find the topics to subscribe to. 

**To Reproduce**
1. Create a Consumer like below
```
client
      .newConsumer(Schema.BYTES)
      .topicsPattern(Pattern.compile("persistent://\w+/Test/Test1"))
      .subscribe()
```
2. Observe `java.lang.IllegalArgumentException: Invalid named entity: \w+` is thrown 

**Expected behavior**
Given a wildcard in a tenant or namespace I would expect the code to look up all tenants/namespaces that match and then lookup all topics to subscribe to.

**Proposed Solution**
1. LookupService implementations should implement a method like `getNamespaces()` that will return all tenant/namespaces. 
2. `PulsarClientImpl.patternTopicSubscribeAsync` should filter these namespaces based on the regex given
3. For each renaming namespace that matches the given regex `PulsarClientImpl.patternTopicSubscribeAsync` should call `LookupService.getTopicsUnderNamespace`
4. List of topics for each matching namespace should be combined into one list and the consumer should subscribe to all of those topics

### PR DESCRIPTION
**Pull request by rmn36**
Fixes #7846

### Motivation
Given a wildcard in a tenant or namespace I would expect the code to look up all tenants/namespaces that match and then lookup all topics to subscribe to.

### Verifying this change

- Updated PatternMultiTopicsConsumerImplTest

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): (yes / no)
  - The public API: (**yes** / no)
  - The schema: (yes / **no** / don't know)
  - The default values of configurations: (yes / **no**)
  - The wire protocol: (yes / **no**)
  - The rest endpoints: (**yes** / no)
  - The admin cli options: (yes / no)
  - Anything that affects deployment: (yes / **no** / don't know)

### Documentation

  - Does this pull request introduce a new feature? (**yes** / no)
  - If a feature is not documented yet in this PR, please create a followup issue for adding the documentation

### CODE REVIEW

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/NamespacesBase.java [hunk 1]
```diff
@@ -30,17 +30,12 @@
 import java.lang.reflect.Field;
 import java.net.URI;
 import java.net.URL;
-import java.util.Collections;
-import java.util.List;
-import java.util.Map;
+import java.util.*;
```
Comment (jiazhai) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/NamespacesBase.java:33]: Please avoid using ".*"

**Review by jiazhai — APPROVED** [review]

**Comment by jiazhai** [comment]
@rmn36 Seems some of the error in ci is related to this change.
e.g. cpp ci test:
```
/pulsar/pulsar-client-cpp/lib/Commands.cc: In static member function 'static std::__cxx11::string pulsar::Commands::messageType(pulsar::proto::BaseCommand_Type)':
/pulsar/pulsar-client-cpp/lib/Commands.cc:464:12: error: enumeration value 'BaseCommand_Type_GET_NAMESPACES_BY_REGEX' not handled in switch [-Werror=switch]
     switch (type) {
            ^
/pulsar/pulsar-client-cpp/lib/Commands.cc:464:12: error: enumeration value 'BaseCommand_Type_GET_NAMESPACES_BY_REGEX_RESPONSE' not handled in switch [-Werror=switch]
[  9%] Building CXX object lib/CMakeFiles/pulsarStatic.dir/ClientConnection.cc.o
[  9%] Building CXX object lib/CMakeFiles/pulsarStatic.dir/TopicName.cc.o
```

**Comment by jiazhai** [comment]
@rmn36 Would you please help rebase the code with master, and get the ci fixed?

**Comment by codelipenghui** [comment]
@rmn36 Would you please help resolve the conflicts?

**Comment by codelipenghui** [comment]
move to 2.8.0 first

**Comment by Anonymitaet** [comment]
@rmn36 thanks for your great work. Could you please help add docs accordingly? Thanks

**Comment by Anonymitaet** [comment]
@rmn36 thanks for your work.

1. Any progress for this PR?
2. Could you please help add docs accordingly? Thanks

**Comment by Anonymitaet** [comment]
Hi @rmn36 any progress for the code and doc? thanks

**Comment by siaron** [comment]
Is there any update? We really need this feature.

**Comment by TonyYANOqton** [comment]
Hi, anybody knows when this feature is expected to be released?  We really need this feature too.

**Comment by tisonkun** [comment]
Closed as stale and conflict. Please rebase and resubmit the patch if it's still relevant.

**Comment by tisonkun** [comment]
@TonyYANOqton This seems a new feature that can require a PIP to proceed. I move the original issue to the discussion forum https://github.com/apache/pulsar/discussions/18862 since we don't have a consensus on whether this is the right way yet.

As you can see, this patch conflicts a lot of files and it requires significant effort to rebase.

**Pull request closed without merging**